File: web/src/model/registry/elements/busLogicInput.ts
/**
 * The bus logic input (BusLogicInputElm.java, XML type "bli"): an N-bit wide
 * logic driver. One part holds a word; bit i of it drives post i with hiV or
 * loV, and every post sits on the anchor coordinate carrying its own bus bit,
 * so a splitter's bus side or a bus wire meets all N bits at once. Clicking
 * cycles the word 0..2^N-1, upstream's `toggle()` (BusLogicInputElm.java:
 * 116-120).
 *
 * Upstream saves this class only in the XML format (its text dump type is 0),
 * so the port assigns dump code 435, free upstream and beside the other
 * port-assigned XML-era codes (the instruction display's 434).
 */

import { canvasFont, currentDots, elementLength, interp, line, limbColor, voltageColor } from '../../../render/draw';
import type { CircuitElement, DrawContext, ElementDef } from '../../types';

/** The bit count the engine accepts: truncated and clamped to the 2..32 its
 *  constructor enforces (bus_logic_input.rs). */
export function normalizeBusLogicInputWidth(value: number): number {
  if (!Number.isFinite(value)) return 4;
  return Math.min(32, Math.max(2, Math.trunc(value)));
}

function busWidth(e: CircuitElement): number {
  return normalizeBusLogicInputWidth(e.params.busWidth ?? 4);
}

/** The live word: the toggled session state wins over the file token, the
 *  same precedence every interactive part uses. */
export function busInputValue(e: CircuitElement): number {
  const raw = e.state ?? e.params.value ?? 0;
  const max = 2 ** Math.min(31, busWidth(e));
  return Math.max(0, Math.trunc(raw)) % max;
}

function drawBusLogicInput(g: DrawContext, e: CircuitElement): void {
  const p1 = { x: e.x1, y: e.y1 };
  const p2 = { x: e.x2, y: e.y2 };
  const dn = Math.max(1, elementLength(e));
  // A short thick lead from the anchor (drawThickLine at weight 5,
  // BusLogicInputElm.java:100), then the bold value at the drag end.
  const lead1 = interp(p1, p2, 1 - 12 / dn);
  line(g, p1, lead1, voltageColor(g, g.voltages[0]), 5, 'round');
  currentDots(g, lead1, p1, g.current);
  g.ctx.fillStyle = limbColor(g, g.theme.text);
  g.ctx.font = `bold ${canvasFont(20)}`;
  g.ctx.textAlign = 'center';
  g.ctx.textBaseline = 'middle';
  g.ctx.fillText(String(busInputValue(e)), p2.x, p2.y);
}

export const BUS_LOGIC_INPUT_DEF: ElementDef = {
  kind: 'busLogicInput',
  label: 'Bus logic input',
  category: 'Logic',
  dumpCode: '435',
  postCount: 4,  // the busWidth(4) default, for the fresh-part fallback
  postCountOf: (e) => busWidth(e),
  posts: (e) => {
    // Every pin shares the anchor, one per bit (getPost(n) = new Point(x, y,
    // n), BusLogicInputElm.java:61-63); the engine tags them with their bit.
    const n = busWidth(e);
    return Array.from({ length: n }, () => ({ x: e.x1, y: e.y1 }));
  },
  draggablePosts: 2,  // the value end is a control point, not a terminal
  noDiagonal: true,
  interactive: true,
  defaults: { busWidth: 4, hiV: 5, loV: 0 },
  parse: (t, e) => {
    // The port's own stream: width, word, then the two levels.
    const bw = Number(t[0]);
    if (Number.isFinite(bw)) e.params.busWidth = normalizeBusLogicInputWidth(bw);
    const va = Number(t[1]);
    if (Number.isFinite(va)) e.params.value = va;
    const hi = Number(t[2]);
    if (Number.isFinite(hi)) e.params.hiV = hi;
    const lo = Number(t[3]);
    if (Number.isFinite(lo)) e.params.loV = lo;
    e.state = e.params.value ?? 0;
  },
  // The word written is the live toggled state when present, falling back to
  // the file token: every other interactive part saves its session state this
  // way, and a click between save points would otherwise be lost.
  dump: (e) => [
    busWidth(e),
    busInputValue(e),
    e.params.hiV ?? 5,
    e.params.loV ?? 0,
  ],
  fields: [
    { name: 'busWidth', label: 'Bus Width', min: 2, max: 32, integer: true },
    { name: 'value', label: 'Value', integer: true },
    { name: 'hiV', label: 'High logic voltage', unit: 'V' },
    { name: 'loV', label: 'Low voltage', unit: 'V' },
  ],
  draw: drawBusLogicInput,
};

File: engine/core/src/elements/bus_logic_input.rs
//! Bus logic input (BusLogicInputElm.java, XML type "bli").
//!
//! An N-bit wide logic input: one part drives N independent nodes, bit i of
//! its `value` becoming hiV or loV on post i. Every post sits at the same
//! coordinate carrying its own bus bit (`getPost(n) = new Point(x, y, n)`,
//! BusLogicInputElm.java:61-63), so a bus splitter's bus side or a bus wire
//! meets all N bits at once.
//!
//! Upstream saves this class only in the XML format (`getDumpType` is 0), so
//! the port assigns dump code 435: free upstream, and next to the other
//! port-assigned XML-era codes (the instruction display's 434).

use crate::element::{Base, Element, SimCtx};
use crate::spec::ElementSpec;
use crate::stamp::{Stamper, GROUND};

pub struct BusLogicInput {
    base: Base,
    /// Driven levels.
    hi_v: f64,
    lo_v: f64,
    /// The driven word; bit i drives post i.
    value: u32,
    width: usize,
    /// Per-bit source currents from the last solved step, the port of
    /// upstream's `currents[]` that setCurrent fills
    /// (BusLogicInputElm.java:73-79).
    per_bit_currents: Vec<f64>,
}

impl BusLogicInput {
    pub fn new(spec: &ElementSpec) -> Self {
        // The edit dialog rejects fewer than 2 bits
        // (BusLogicInputElm.java:155-160); the clamp also bounds the post and
        // source count against a corrupt token.
        let width = (spec.param("busWidth", 4.0) as usize).clamp(2, 32);
        Self {
            base: Base::with_posts(width),
            hi_v: spec.param("hiV", 5.0),
            lo_v: spec.param("loV", 0.0),
            value: spec.param("value", 0.0) as u32,
            width,
            per_bit_currents: vec![0.0; width],
        }
    }

    fn bit_voltage(&self, bit: usize) -> f64 {
        if self.value & (1 << bit) != 0 {
            self.hi_v
        } else {
            self.lo_v
        }
    }
}

impl Element for BusLogicInput {
    fn kind(&self) -> &'static str {
        "busLogicInput"
    }
    fn base(&self) -> &Base {
        &self.base
    }
    fn base_mut(&mut self) -> &mut Base {
        &mut self.base
    }
    fn post_count(&self) -> usize {
        self.width
    }
    /// Every post shares the anchor coordinate and each carries its own bit
    /// (`getPost(n) = new Point(x, y, n)`, BusLogicInputElm.java:61-63), so
    /// the bit tags are what keep the pins apart.
    fn post_bus_z(&self, post: usize) -> usize {
        post
    }
    /// One source per bit, spanning ground to that bit's post
    /// (getVoltageSourceCount and stamp, BusLogicInputElm.java:59, :122-127).
    fn voltage_source_count(&self) -> usize {
        self.width
    }
    fn voltage_source_nodes(&self, k: usize) -> (usize, usize) {
        (GROUND, self.base.nodes[k])
    }
    /// Each post is a voltage-source output; no two posts couple
    /// (isWireEquivalent false, BusLogicInputElm.java:131).
    fn connects(&self, _a: usize, _b: usize) -> bool {
        false
    }
    /// The CAP_V walk must be able to cross these sources, exactly like the
    /// rail and the single-bit logic input.
    fn is_voltage_source(&self) -> bool {
        true
    }
    fn stamp(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        for k in 0..self.width {
            s.voltage_source(GROUND, self.base.nodes[k], self.base.vs_base + k, 0.0);
        }
    }
    fn do_step(&mut self, _ctx: &SimCtx, s: &mut Stamper) {
        for k in 0..self.width {
            s.voltage_source_value(self.base.vs_base + k, self.bit_voltage(k));
        }
    }
    fn calculate_current(&mut self, _ctx: &SimCtx) {
        // The engine hands the solved source currents back in vs_currents;
        // keeping them per bit is what makes the wire-current recovery see
        // each bit separately.
        self.per_bit_currents
            .copy_from_slice(&self.base.vs_currents);
        self.base.current = self.per_bit_currents.iter().sum();
    }
    fn current_into_node(&self, post: usize) -> f64 {
        // getCurrentIntoNode returns currents[n] per bit
        // (BusLogicInputElm.java:81-83).
        self.per_bit_currents.get(post).copied().unwrap_or(0.0)
    }
    /// The readout shows the post-0 level like the single-bit input does.
    fn voltage_diff(&self) -> f64 {
        self.base.volts.first().copied().unwrap_or(0.0)
    }
    /// Clicking cycles the value through 0..2^width-1 (toggle,
    /// BusLogicInputElm.java:116-120); the frontend wraps and sends the next
    /// word.
    fn set_state(&mut self, state: i32) -> bool {
        self.value = state.max(0) as u32;
        true
    }
    fn set_param(&mut self, name: &str, value: f64) -> bool {
        match name {
            "hiV" => self.hi_v = value,
            "loV" => self.lo_v = value,
            "value" => self.value = value as u32,
            _ => return false,
        }
        true
    }
    fn reset(&mut self) {
        self.base.reset();
        self.per_bit_currents.iter_mut().for_each(|c| *c = 0.0);
    }
}
